### DISCUSSION TITLE
How to advance through vertex buffer across instances?

### DISCUSSION THREAD

**Discussion by smolck** [gfx-rs/wgpu#2016] [Q&A] [answered]
Not quite sure how to phrase this question succinctly, so apologies if the title here doesn't quite make sense.

Basically, let's say I have two vertex buffers, the first with `VertexStepMode::Instance` and the second with `VertexStepMode::Vertex`, and I'm rendering a series of quads in a grid. The first buffer contains position etc. information for each of those quads, and is instanced since I only want to advance that once per 6 vertices; the second one contains texture coordinate information used to sample from a texture for each quad, and should advance each vertex.

The problem I've run into, however, is that afaict only the first 6 `vec2<f32>`s (being the texture coordinates used for each vertex, as mentioned earlier) of the second buffer are ever used in rendering. Or put another way, where the data is taken from in the second buffer seems to be reset each instance, so it starts back at the beginning of the buffer for each instance instead of continuing further into the buffer, as I'd like it to.

For context, my draw call is essentially this: `render_pass.draw(0..6, 0..num_instances)`. The 0..6 would make sense to cause this behavior I guess, however, is there a way of changing it? So that it will continually advance through the second vertex buffer across instances instead of repeating at the start for each instance?

If any of this is unclear, let me know and hopefully I can explain it better, or if not I can try to provide a minimal code example to better showcase what I'm running into here.

**Comment by kvark** [comment, marked as answer]
Graphics APIs do not offer that. If you want the advancement to be done for each vertex *and* also for instances, then you'd need to make this a non-instanced draw call. As for the buffer you are currently using as instanced, there are at least 2 choices:
  1. make it so it duplicates the info for each vertex in a quad. It's a waste, but if you don't draw that many quads it could be OK.
  2. bind this buffer as a read-only storage buffer instead of a vertex buffer, and load data from it via something like `gl_VertexIndex / 6` index. This would mean the buffer is exactly what you have today, and it's advanced once per 6 vertices.

**Reply by cwfitzgerald** [reply]
The secondary option has the benefit of having a single large instance, which gpus are much happier with.

**Reply by smolck** [reply]
Thank you for the answer! That makes sense; I'll see about implementing the second option and hopefully get something working.

> The secondary option has the benefit of having a single large instance, which gpus are much happier with.

@cwfitzgerald TIL! Why is this/anywhere I can read up on this?

**Reply by cwfitzgerald** [reply]
I unfortunately don't have a good source besides graphics hearsay and empirical testing. I never quite figured out _why_ it is slower. Intel/AMD struggled with instancing small things, Nvidia struggled with instancing small things, but only when doing indexed rendering. You often hear this advice as "don't instance quads", for the same reason.

**Reply by CatCode79** [reply]
@smolck Try this GpuGems article, I think it can help you a little:
https://developer.nvidia.com/gpugems/gpugems2/part-i-geometric-complexity/chapter-3-inside-geometry-instancing
